### ISSUE TITLE
Differential drive robot (ignoring directional contraints during physics substeps)

### ISSUE THREAD

**Issue by Zartris** [proroklab/VectorizedMultiAgentSimulator#46]
I am trying to implement a differential drive robot. 
The velocity controller gives an excellent example of how to change the action output of a network into the forces that the simulator is expecting.
The "problem" I'm running into is the _integrate_state function in the core.py.
The code expects the orientation of the entity and the forces (velocity of the robot) to be independent. However, when applied to a differential-driven agent, rotating means we are also rotating our velocity vector.
Long story short, the simulation is not working well when you have a directional constraint on your movement, like a differential drive (as it ignores the constraints during substeps and is only applied before every step).

It can be hacked quickly enough for my project, so there is no need for an immediate solution. Still, as you explicitly mention the differential drive dynamic model on the front page of the repo, it might be something you should be aware of.

**Comment by matteobettini** [comment]
Thanks so much for this!

Yes having a diff drive model is actually something we are extremely interested in.
And my idea was that it can be done in the preprocessing like the velocity controller, with a component that takes forward and rotational vel and transforms those in forces and torque.

But I see that this might not be possible with the current core.

Do you think you have a way to solve this so that we can integrate it in vmas and is compatible with the current dynamics?

I would be happy to review a PR if you want to post one and to merge your solution.

**Comment by Zartris** [comment]
I am still working on the controller in the preprocessing ( or rather testing if it works as I wanted it to), but I am sure it can be implemented nicely into VMAS. I don't have a good solution currently, and I am soon going to ICRA in London, so it will not be submitted before I'm back. Any plans on joining ICRA this year? It should be close to your lab.

**Comment by matteobettini** [comment]
Yes we will be at ICRA! 
We can meet up, you can definitely find me on monday [here](https://sites.google.com/cam.ac.uk/multirobotlearningworkshopicra/home) or just drop me a message to organize.

**Comment by Zartris** [comment]
Great, I am in the same workshop, so I hope we can have a chance to talk a bit.

**Comment by matteobettini** [comment]
i am working on a diff drive class with euler and runge kutta which should not have too many problems with substepping.

I ll post the PR in a bit and I'll link it here

**Comment by Zartris** [comment]
Great! I'll take a look at it when you post it.

**Comment by matteobettini** [comment]
check out #47

**Comment by Zartris** [comment]
Hey @matteobettini, sorry for the late reply, as I had a busy Friday.

So far, it looks very elegant, and I have yet to run into any problems.
Let's keep this closed, and I will open new tickets when I gain further knowledge of your simulator.
I'm more or less settled with using your simulator for my next project, so I hope you don't mind me opening tickets whenever I find something to improve or if I do not understand your code.
There are some performance issues here and there, but that is just the nature of Python and the design choices of the simulator.
Usually, I fix my own small problem as some of them are very project dependent and will take away from the generality of the simulator, but I also like to let you know the problems I have as maybe others are running into similar issues. 

See you at Icra.
